### PR TITLE
#[drink::contract_bundle_provider] macro

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
Follow-up to #70 and #68.

We introduce a new macro `#[drink::contract_bundle_provider]` which provides a convenient way of obtaining contract bundles in drink testing.

This bundling provision wasn't done within `#[drink:test]` macro so as to avoid generating enums and implementations per-testcase. Now the user has to (although they can still get bytes/transcoders manually or use `ContractBundle` API) add one new type to the test module:
```rust
#[drink::contract_bundle_provider]
enum BundleProvider {}
```

which will be expanded to:

```rust
enum BundleProvider {
    Flipper,
    Counter,
}
impl BundleProvider {
    pub fn local() -> ::drink::DrinkResult<::drink::ContractBundle> {
        ::drink::ContractBundle::load(
            "/home/piomiko/Desktop/drink/examples/flipper/target/ink/flipper.contract",
        )
    }
    pub fn bundle(self) -> ::drink::DrinkResult<::drink::ContractBundle> {
        match self {
            BundleProvider::Flipper => ::drink::ContractBundle::load(
                "/home/piomiko/Desktop/drink/examples/flipper/target/ink/flipper.contract",
            ),
            BundleProvider::Counter => ::drink::ContractBundle::load(
                "/home/piomiko/Desktop/drink/examples/counter/target/ink/counter.contract",
            ),
        }
    }
}
```
(assuming that we are using it within `flipper` project and we have `counter` as a contract dependency in `Cargo.toml`)

### CODE REVIEW

File: examples/flipper/lib.rs
Comment (deuszx) [inline examples/flipper/lib.rs:62]: This will get red underlined squiggles?

Comment (pmikolajczyk41) [reply examples/flipper/lib.rs:62]: it depends on the IDE - mine (Rust Rover) is able to resolve it and even give hints on variants / methods :sunglasses:

it's quite a lightweight macro and thus some decent Rust analyzers should get it

File: examples/mocking/lib.rs
Comment (deuszx) [inline examples/mocking/lib.rs:52]: Not a fan of the DX - it's too magical for me. Personally I prefer the "in your face" approaches.

Comment (pmikolajczyk41) [reply examples/mocking/lib.rs:52]: you can't get 'in your face' approach baking such info in the type system... I can see 3 options here:
 - drink provides no support for bundles, i.e. you have to use `ContractBundle` API with string literals, and its your responsibility to care for paths
 - drink provides e.g. mapping from contract name / manifest file to a `ContractBundle` as additional testcase argument (so your tests actually get one argument from the sky), and you depend on string identifiers / paths, for which you are responsible
 - drink provides a typed source of bundles (this PR)

Comment (deuszx) [reply examples/mocking/lib.rs:52]: The thing with this PR is that I type in code that means nothing if you look at it. What about, sth like this:
```rust
#[drink::test(
  flipper = "path/to/flipper",
  counter = "path/to/counter",
  ...
)]
fn test(bundles: HashMap<String, ContractBundle>) {
 // flipper and counter available here
}
```

Which is basically what you propose in 2)?

Comment (pmikolajczyk41) [reply examples/mocking/lib.rs:52]: how does these paths relate to dependencies in `Cargo.toml`? it'd be not cool to specify them for every testcase (actually, macro already knows all the paths)

---

> means nothing if you look at it

means more if you know what it does - every information is implicit (like contract names)

Comment (deuszx) [reply examples/mocking/lib.rs:52]: > how does these paths relate to dependencies in Cargo.toml? it'd be not cool to specify them for every testcase (actually, macro already knows all the paths)

Hmm, good point. It might overwrite deps in `Cargo.toml`. But also it allows for adding ones that are not defined there. 

> means more if you know what it does - every information is implicit (like contract names)

Exactly. And this is what I don't like about it. 

I'm not gonna block the PR over it ofc.

File: drink/test-macro/src/lib.rs
Comment (deuszx) [inline drink/test-macro/src/lib.rs:87]:
```suggestion
/// Otherwise, the enum will receive a method `bundle(self)` that returns the contract bundle
```
?

Comment (pmikolajczyk41) [reply drink/test-macro/src/lib.rs:87]: nope, it's the `local` function which is an addition - `bundle` must be present to expose all available bundles

File: drink/test-macro/src/bundle_providing.rs
Comment (piotrMocz) [inline drink/test-macro/src/bundle_providing.rs]: This is extremely nitpicky but I'd prefer `s/providing/provision/g` :smile:

**Review by piotrMocz — APPROVED** [review]
I think this is great! I'd say it still may have the potential to make it even more user-friendly but we can definitely discuss this after merging of this one (which I think is already a great step)

File: drink/test-macro/src/bundle_providing.rs
Comment (pmikolajczyk41) [reply drink/test-macro/src/bundle_providing.rs]: sure, done!

**Pull request merged**
